### DISCUSSION TITLE
Initialize kart on pre existing PostGIS database

### DISCUSSION THREAD

**Discussion by SrNetoChan** [koordinates/kart#681] [Q&A] [answered]
Not sure if this is the right place to ask this.

Having a pre-existent database in PostgreSQL/PostGIS, already containing data is it possible to initialize a kart working copy in it and add only some of the pre-existing tables to kart versioning system?

Would that make sense?

I am looking for a convenient way to keep working normally in our PostGIS database, and use kart only on necessary tables.

**Comment by rcoup** [comment, marked as answer]
Yes, definitely makes sense.

Currently there's no way to convert a pre-existing table into a Kart table _in place_, but kart & non-kart tables can happily coexist in the same database — if you add the `kart` schema into your search path (eg: `ALTER DATABASE mydb SET search_path TO  "$user", public, kart;`) then a query will find both kart & non-kart tables without needing the schema qualifier (first in search_path is preferred).

So, you can convert a table (`mytable`) that's in the `public` schema into a Kart table along these lines, if you don't have other related tables referring to fields in `mytable`. Practise this on a database copy first!
1. import `mytable` into the kart repository, so it gets checked out into your database as `kart.mytable`
2. alter the search path as above
3. drop the original mytable (`public.mytable`)
4. any queries referring to `mytable` will now use `kart.mytable` instead.

Related to https://github.com/koordinates/kart/issues/249

**Reply by marcathian** [reply]
How do I get kart to initialize a working copy in postgis? 
I have been trying: kart init --workingcopy = postgresql://[user[:password]@]host[:port]/dbname/dbschema
But It only makes a tree of folders within the current directy.  All the parameters typed into the URI are used to name the series of folders. e.g:

-- postgresql:
    --user:password@host:port
       --dbname
          --dbschema
No connection is established with the database

Thank you for your assistance

**Reply by craigds** [reply]
@marcathian it looks like you've got extra spaces in your command there. Kart is probably parsing it like this:

* `--workingcopy =` - put the workingcopy in a file called `=`
* `postgresql://[user[:password]@]host[:port]/dbname/dbschema` - put the repo itself in a folder called `postgresql://[user[:password]@]host[:port]/dbname/dbschema`

If you remove the spaces between those it should work better:

`kart init --workingcopy=postgresql://[user[:password]@]host[:port]/dbname/dbschema`

**Reply by marcathian** [reply]
Thank you very much.  I was able to move on and create the postGIS repo and import data into it.  However I am not able to load the data into a QGIS project using the QGIS plugin. I get the error when using both the GeoPackage and PostGIS working copy types.  However the data is still loaded from the GeoPackage working copy type.


[Kart_qgis_plugin_error-on_addToProject.txt](https://github.com/koordinates/kart/files/9796183/Kart_qgis_plugin_error-on_addToProject.txt)

**Reply by pfw** [reply]
The error you are seeing in QGIS is a known one we are actively working on and should have a status update for you within the next few days.

**Reply by SrNetoChan** [reply]
For my specific user case, I needed to put my working copy on the public schema. I am failing to do it. Mostly because the schema to init kart needs to be empty. Then if I try to import data, it misses postgis. If I install postgis, it says the working copy is out of sync.  Can any of you think of a way to use public as the kart working copy?

Thanks!!

**Reply by rcoup** [reply]
@SrNetoChan I'm keen to understand your requirement to use the `public` schema — what's the practical difference between that and doing `ALTER DATABASE mydb SET search_path TO public,kart;`? With that in place you can just do `SELECT * FROM mykartdataset;` the same way as you could if a table is in `public`.

wrt Postgis, you can install _it_ in another schema too: `CREATE EXTENSION postgis SCHEMA postgis;`. You then might want `ALTER DATABASE mydb SET search_path TO public,postgis;`

📖  [PostgreSQL docs on the schema search path](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH)

**Reply by SrNetoChan** [reply]
Well, I am working with a national (Portugal) data model that was created to store and share topographic cartography. Unfortunately, the data model was built around the public schema. All the qgis project and plugins are expecting the public schema. And once ready, the database of a particular location needs to be sent to the national cartographic authority for revision. If the tables are not in the public schema, it comes back saying it's not complying with data model.

Nevertheless, installing postgis in a separate schema might just work. I will let you know if it did!
